src/codegen/generators/expression_generator.rs

```diff
@@ -504,9 +504,16 @@ impl<'ink, 'b> ExpressionCodeGenerator<'ink, 'b> {
         // if the function is builtin, generate a basic value enum for it
         if let Some(builtin) = self.index.get_builtin_function(implementation_name) {
             // adr, ref, etc.
-            return builtin
-                .codegen(self, parameters_list.as_slice(), operator.get_location())
-                .map(ExpressionValue::RValue);
+            return builtin.codegen(self, parameters_list.as_slice(), operator.get_location()).map(|it| {
```

Comment:
can you explain a bit further in what cases we need an LValue here? I dont really understand. Previously we always returned an RValue - also for REF and ADR. 
now we keep returning an Rvalue for REF and ADR, except it is a builtin function except REF/ADR that returns a pointer. Maybe we can talk about this in person - probably I'm missing something

Reply:
After Ghaith and i looked into the issue, we decided for aggregate return types to pass a pointer to SEL
not the value itself
i think in this case the returned values should be LValues
i refactored the builtins so they return an ExpressionValue instead
if you think there is a better solution im open to discuss 👍🏼

Reply:
I really like your approach - we should really carefully consider which of these functions return an LValue and which return an RValue ... so thats great!

I found some issue with the MUX, which I believe is returning an LValue, bit once I tested it with strings, the annotator started to do funny things - it is really strange that the MUX works with structs and arrays, as soon as it is strings, it starts to annotate the mux-selector (K) with a STRING type-hint. I added a resolver-test as well, this one shows the effect - unfortunately I'm a bit buzy so I hand this back to you. Maybe somebody did a fix not long ago that causes this issue and knows quickly where to look exactly?!?!?

so `MUX(1, 'a', 'b', 'c');` will magically annotate the 1 with Type=DINT and TypeHint=STRING :thinking:

Reply:
Dont worry too much about the LValue and RValue stuff, it is not really obvious and it is really a compiler-internal concept - I also did not introduce it very well. I think we should discuss the whole thing again with the whole team - I'm pretty sure there is lots of confusion happening.

Reply:
the problem leading to this false annotation was in generics::update_generic_function_parameters
we need to annotate the generic parameters with the derived generic type
but if we found a variadic parameter, we annotated all passed parameters with the same type, we didn't filter to annotate just the variadics